execution-engine/src/native_modules/darknet_inference.rs

```diff
@@ -0,0 +1,202 @@
+//! A native module for ML inference on Darknet.
+//! Takes an input image, feeds it to the model and outputs a list of detected
+//! objects.
+//!
+//! ## Authors
+//!
+//! The Veracruz Development Team.
+//!
+//! ## Licensing and copyright notice
+//!
+//! See the `LICENSE_MIT.markdown` file in the Veracruz root directory for
+//! information on licensing and copyright.
+
+use crate::fs::{FileSystem, FileSystemResult};
+use crate::native_modules::common::Service;
+use darknet::{BBox, Detection, Image, Network};
+use lazy_static::*;
+use serde::Deserialize;
+use std::cmp::Ordering;
+use std::env::{current_dir, set_current_dir};
+use std::fmt::Write as _;
+use std::fs::{create_dir, create_dir_all, read_to_string, remove_dir_all, File};
+use std::io::Write;
+use std::path::{Path, PathBuf};
+use wasi_types::ErrNo;
+
+lazy_static! {
+    // Current directory before moving to the module's directory
+    static ref WORK_DIR: PathBuf = current_dir().unwrap_or(PathBuf::from(""));
+
+    // The native module executes in a dedicated directory to mitigate data
+    // leaks
+    static ref MODULE_DIR: PathBuf = WORK_DIR.clone().join("darknet_inference_service");
+}
+
+/// Module's API.
+#[derive(Deserialize, Debug)]
+pub(crate) struct DarknetInferenceService {
+    /// Path to the input to be fed to the network.
+    input_path: PathBuf,
+    /// Path to the model's configuration.
+    cfg_path: PathBuf,
+    /// Path to the actual model (weights).
+    model_path: PathBuf,
+    /// Path to the labels file containing all the objects that can be detected.
+    labels_path: PathBuf,
+    /// Path to the output file containing the result of the prediction.
+    output_path: PathBuf,
+}
+
+impl Service for DarknetInferenceService {
+    /// Return the name of this service
+    fn name(&self) -> &str {
+        "Darknet Inference Service"
+    }
+
+    /// Triggers the service. The details of the service can be found in
+    /// function `infer`.
+    /// Here is the enter point. It also erases the state unconditionally
+    /// afterwards.
+    fn serve(&mut self, fs: &mut FileSystem, _input: &[u8]) -> FileSystemResult<()> {
+        // when reaching here, the `input` bytes are already parsed.
+
+        // Prepare module's directory
+        create_dir(MODULE_DIR.as_path()).map_err(|_| ErrNo::Access)?;
```

Comment:
`create_dir` can fail for reasons other than lack of access, right?

Reply:
Yes. I could return `Canceled`, since that seems to be the default error for native modules (cf. AES modules)